cmd/gateway/s3/gateway-s3.go

```diff
@@ -289,7 +289,7 @@ func (l *s3Objects) GetObjectInfo(bucket string, object string) (objInfo minio.O
 
 // PutObject creates a new object with the incoming data,
 func (l *s3Objects) PutObject(bucket string, object string, data *hash.Reader, metadata map[string]string) (objInfo minio.ObjectInfo, err error) {
-	oi, err := l.Client.PutObject(bucket, object, data, data.Size(), data.MD5HexString(), data.SHA256HexString(), minio.ToMinioClientMetadata(metadata))
+	oi, err := l.Client.PutObject(bucket, object, data, data.Size(), data.MD5Base64String(), data.SHA256HexString(), minio.ToMinioClientMetadata(metadata))
```

Comment:
this is why previously md5sum and sha256 were of the type `[]byte` so that it is hex/base64 representation independent and bugs like this are avoided .... we need to revert back to that api in minio-go. however we can take this patch in.

Reply:
We shouldnt send byte that is incorrect. We should give the right encoding this PR is correct. The problem is that we are doing double validation that can be fixed when we truly send the right values to the gateway which we didn't so this PR fixes.

For example let's say we remove hashReader for gateway completely we will be sending the stream directly, if the headers have Content-MD5 we should preserve it and send it directly.

Reply:
> so that it is hex/base64 representation independent and bugs like this are avoided

In-fact bugs like this are good it exposes the flow, we should always send the right encoding to a low level API and not expect it to convert to a right encoding.  The encoded values are already there the problem we have is double validation in gateway.  A stream is not necessary to be re-validated since all backends have a way to validate the incoming stream. 

For example 
- https://github.com/minio/minio/issues/4995
- https://github.com/minio/minio/issues/4994

We are not using the gcs/azure features to validate `Content-Md5`

Reply:
if an API takes input in canonical form then programmer bugs like this can be avoided.

> For example let's say we remove hashReader for gateway completely we will be sending the stream directly, if the headers have Content-MD5 we should preserve it and send it directly.

you are thinking about `core.PutObject` API from what is convenient for the gateway i.e gateway can directly pass the `Content-Md5` to PutObject(). But if you think of `core.PutObject` as an API for applications `[]byte` is the canonical way to represent a hash. i.e developers who are using core APIs directly will now have to be careful regarding what type of encoding they do for md5 and sha256.

Reply:
Canonical form is not []byte, its the right encoding for HTTP headers is string. The interpretation you are putting forward is not correct here.  Look at the PutObject() APIs in aws-sdk-go they require applications to set the right Content-Md5 and they right fully take string not []byte. Because that is how the S3 API is documented.  It would be a misleading API and not representing the right requirements of S3 API.

Reply:
In-fact if we used aws-sdk-go this argument wouldn't exist.  All aws-sdk's right fully document this correctly, neither of them take a byte representation.  Here API is documenting what is required from the user we also named it correctly If someone reads our godoc. 

https://godoc.org/github.com/minio/minio-go#Core.PutObject

```
func (c Core) PutObject(bucket, object string, data io.Reader, size int64, md5Base64, sha256Hex string, metadata map[string]string) (ObjectInfo, error)
```

```
func (c Core) PutObjectPart(bucket, object, uploadID string, partID int, data io.Reader, size int64, md5Base64, sha256Hex string) (ObjectPart, error)
```

One problem is that disconnection between exposed PutObject() v/s core PutObject() i would prefer a consolidation effort where PutObject core API takes *PutObjectOptions which in-fact has a method SetContentMD5() 

http://docs.aws.amazon.com/sdk-for-go/api/service/s3/#PutObjectInput.SetContentMD5

Reply:
> Canonical form is not []byte
in Go `[]byte` is the canonical form for md5 sum, infact it is `[16]byte` which provides even better typing.

>  Look at the PutObject() APIs in aws-sdk-go
Wouldn't lookup to aws-sdk-go for API design

Let me check with AB what he thinks.

Reply:
Well not just aws-sdk-go it's for all SDKs. A canonical form term should be associated to the API/Protocol in question not the language IMO.

PutObjectOptions for example is the right place for these values it already has other values and we can extend that.